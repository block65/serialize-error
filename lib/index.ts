import { CustomError } from '@block65/custom-error';
import { serializeError as serialize } from 'serialize-error';

export interface SerializedError {
  name: string;
  message: string;
  stack: string[];
  sensitive?: boolean;
  statusCode?: number;
  previous?: SerializedError[];

  [key: string]: unknown;
}

function flattenPreviousErrors(
  err: Error | CustomError,
  accum: Error[] = [],
): Error[] {
  if ('previous' in err && err.previous) {
    const { previous, ...rest } = err;
    return flattenPreviousErrors(previous, [...accum, rest]);
  }
  return [...accum, err];
}

function prepareStack(stack?: string): string[] {
  return (stack || '')
    .split('\n')
    .map((frame): string => frame.trim())
    .slice(1);
}

export function serializeError(
  err: unknown | Error | CustomError,
): SerializedError {
  if (err instanceof CustomError) {
    const previousErrors =
      'previous' in err && err.previous
        ? flattenPreviousErrors(err.previous)
        : [];

    return {
      ...serialize(err),
      message: err.message,
      name: err.name,
      ...('statusCode' in err ? { statusCode: err.statusCode } : {}),
      stack: prepareStack(err.stack),
      ...('sensitive' in err ? { sensitive: err.sensitive } : {}),
      previous: previousErrors.map(serializeError),
      ...('debug' in err ? { debug: err.debug() } : {}),
    };
  }

  if (err instanceof Error) {
    return {
      ...serialize(err),
      message: err.message || '',
      name: err.name || 'Error',
      stack: prepareStack(err.stack),
    };
  }

  if (typeof err === 'string' || typeof err === 'number') {
    return {
      message: err.toString(),
      name: 'Error',
      stack: prepareStack(Error().stack),
    };
  }

  // Not an error object, maybe primitive or null, undefined
  return {
    name: 'Error',
    message: '',
    stack: prepareStack(Error().stack),
    debug: {
      typeofErr: typeof err,
      err,
    },
  };
}
