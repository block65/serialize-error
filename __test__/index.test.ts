import { CustomError } from '@block65/custom-error';
import { URL } from 'url';
import * as assert from 'assert';
import { serializeError } from '../lib';

function getUrlError(): Error {
  try {
    const url = new URL('/', 'lol');
    throw new Error(`error expected from URL ${url}`);
  } catch (err) {
    return err;
  }
}

function getAssertionError(): Error {
  try {
    assert(false, 'asserting false');
    throw new Error(`error expected`);
  } catch (err) {
    return err;
  }
}

test('AssertionError', () => {
  const err = getAssertionError();

  const result = serializeError(err);

  expect(result).toStrictEqual({
    actual: false,
    code: 'ERR_ASSERTION',
    expected: true,
    generatedMessage: false,
    message: 'asserting false',
    name: 'AssertionError',
    operator: '==',
    previous: [],
    stack: expect.arrayContaining([expect.any(String)]),
  });
});

test('URLError', () => {
  const err = getUrlError();
  const result = serializeError(err);

  expect(result).toStrictEqual({
    code: 'ERR_INVALID_URL',
    input: 'lol',
    message: 'Invalid URL: lol',
    name: 'TypeError',
    previous: [],
    stack: expect.arrayContaining([expect.any(String)]),
  });
});

test('CustomError', () => {
  const previousErr = new CustomError('previous').debug({ woo1: 'woo1' });
  const current = new CustomError('current', previousErr).debug({
    woo2: 'woo2',
  });

  expect(current.previous).toBe(previousErr);

  const result = serializeError(current);

  expect(result).toStrictEqual({
    debug: {
      woo2: 'woo2',
    },
    debugData: {
      woo2: 'woo2',
    },
    message: 'current',
    name: 'CustomError',
    previous: [
      {
        debug: {
          woo1: 'woo1',
        },
        debugData: {
          woo1: 'woo1',
        },
        message: 'previous',
        name: 'CustomError',
        previous: [],
        sensitive: true,
        stack: expect.arrayContaining([expect.any(String)]),
        statusCode: 2,
      },
    ],
    sensitive: true,
    stack: expect.arrayContaining([expect.any(String)]),
    statusCode: 2,
  });
});
