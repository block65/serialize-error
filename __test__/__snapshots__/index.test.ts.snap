// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dont explode on primitives 1`] = `
Object {
  "message": "0",
  "name": "Error",
  "stack": ArrayContaining [
    Any<String>,
  ],
}
`;

exports[`dont explode on primitives 2`] = `
Object {
  "message": "string",
  "name": "Error",
  "stack": ArrayContaining [
    Any<String>,
  ],
}
`;

exports[`dont explode on primitives 3`] = `
Object {
  "debug": Object {
    "err": 0n,
    "typeofErr": "bigint",
  },
  "message": "",
  "name": "Error",
  "stack": ArrayContaining [
    Any<String>,
  ],
}
`;

exports[`dont explode on primitives 4`] = `
Object {
  "debug": Object {
    "err": false,
    "typeofErr": "boolean",
  },
  "message": "",
  "name": "Error",
  "stack": ArrayContaining [
    Any<String>,
  ],
}
`;

exports[`dont explode on primitives 5`] = `
Object {
  "debug": Object {
    "err": Symbol(kaboom),
    "typeofErr": "symbol",
  },
  "message": "",
  "name": "Error",
  "stack": ArrayContaining [
    Any<String>,
  ],
}
`;

exports[`dont explode on voids 1`] = `
Object {
  "debug": Object {
    "err": null,
    "typeofErr": "object",
  },
  "message": "",
  "name": "Error",
  "stack": ArrayContaining [
    Any<String>,
  ],
}
`;

exports[`dont explode on voids 2`] = `
Object {
  "debug": Object {
    "err": undefined,
    "typeofErr": "undefined",
  },
  "message": "",
  "name": "Error",
  "stack": ArrayContaining [
    Any<String>,
  ],
}
`;
